apps/openmw/mwmechanics/aicombat.cpp

```diff
@@ -326,95 +284,66 @@ namespace MWMechanics
         if (currentAction.get())
             currentAction->getCombatRange(rangeAttack, rangeFollow);
 
-        // FIXME: consider moving this stuff to ActionWeapon::getCombatRange
         const ESM::Weapon *weapon = NULL;
-        MWMechanics::WeaponType weaptype = WeapType_None;
-        float weapRange = 1.0f;
+        MWMechanics::WeaponType weapType = MWMechanics::WeaponType::WeapType_None;
+        float spellSpeed = 1.f;
 
-        // Get weapon characteristics
-        if (actorClass.hasInventoryStore(actor))
+        if (dynamic_cast<ActionWeapon*>(currentAction.get()))
         {
-            // TODO: Check equipped weapon and equip a different one if we can't attack with it
-            // (e.g. no ammunition, or wrong type of ammunition equipped, etc. autoEquip is not very smart in this regard))
+            weapon = dynamic_cast<ActionWeapon*>(currentAction.get())->getWeapon();
 
-            //Get weapon speed and range
-            MWWorld::ContainerStoreIterator weaponSlot =
-                MWMechanics::getActiveWeapon(actorClass.getCreatureStats(actor), actorClass.getInventoryStore(actor), &weaptype);
+            if (weapon != NULL)
+                weapType = getWeaponType(static_cast<ESM::Weapon::Type>(weapon->mData.mType));
+            else
+                weapType = WeapType_HandToHand;
 
-            if (weaptype == WeapType_HandToHand)
-            {
-                static float fHandToHandReach =
-                    world->getStore().get<ESM::GameSetting>().find("fHandToHandReach")->getFloat();
-                weapRange = fHandToHandReach;
-            }
-            else if (weaptype != WeapType_PickProbe && weaptype != WeapType_Spell && weaptype != WeapType_None)
+            if ((weapon == NULL && storage.mLastWeapon != "handToHand") 
+                || (weapon != NULL && weapon->mModel != storage.mLastWeapon))
             {
-                // All other WeapTypes are actually weapons, so get<ESM::Weapon> is safe.
-                weapon = weaponSlot->get<ESM::Weapon>()->mBase;
-                weapRange = weapon->mData.mReach;
+                storage.updateWeaponAttackDurations(actor, weapType, weapon == NULL ? 1.f : weapon->mData.mSpeed, ESM::RT_Self);
```

Comment:
The basic idea is to bring AI and character controller closer together, so instead of inspecting the world state and making second guesses about what's going on, the AI can inspect itself.